src/app/plugins/connectfour.plugin.ts

```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
+    const randomMoveCol = moveOptions[Math.floor(Math.random() * moveOptions.length)].col;
+    this._dropPiece(randomMoveCol);
+  }
+
+  // We want the highest difficulty to allow for the least error.
+  private _getNormalizedDifficulty() {
+    return 1 / this._aiDifficulty;
+  }
+
+  // Evaluate the strength of the current board state,
+  // as it relates to Player 2's success.
+  // Return value is the average of all possible moves from this position
+  // Each move is evaluated on a score of [-4, 4]
+  private _evaluate(currentPlayer: number, depth: number) {
+    // If we have reached a win state, then the LAST move won.
+    if (this._checkWin()) {
+      return -4 * currentPlayer;
+    }
+    if (this._checkTie()) {
+      return 0;
+    }
+    // If we reached depth, then evaluate the board.
+    // We use a very simple evaluation: -(longest player one chain)
+    // This causes Lion to simply try to minimize Player 1's success,
+    // While our _checkWin case takes over when it's time to clinch victory.
+    if (depth === this._aiDepth) {
+      return this._longestChainOnBoard(1) - this._longestChainOnBoard(-1);
     }
-    // Subtract one to place piece one slot above obstruction.
+
+    const moves: number[] = [];
+    for (let col = 0; col < this._cols; col++) {
+      if (!this._dropPiece(col, currentPlayer)) {
+        continue;
+      }
+
+      moves.push(this._evaluate(this.currentPlayer * -1, depth + 1));
+      this._removeTopPiece(col);
+    }
+
+    return this._arraySum(moves) / moves.length;
+  }
+
+  private _arraySum(array: number[]): number {
+    return array.reduce((acc, val) => (acc += val));
+  }
+
+  private _dropPiece(col: number, currentPlayer?: number): boolean {
+    let row = this._getFirstObjectInColumn(col);
+    // Place object one row above first object.
     row--;
 
     if (row === -1) {
       return false;
     }
 
-    this._board[row][col] = this.currentPlayer;
+    this._board[row][col] = currentPlayer ? currentPlayer : this.currentPlayer;
 
     return true;
   }
 
-  public checkWin(): boolean {
-    // Returns if a postion is valid and owned by the current player.
-    const validAndOwned: (row: number, col: number) => boolean = (row, col) => {
-      return (
-        row >= 0 &&
-        row < this._rows &&
-        col >= 0 &&
-        col < this._cols &&
-        this._board[row][col] === this.currentPlayer
-      );
-    };
+  private _removeTopPiece(col: number): void {
+    const row = this._getFirstObjectInColumn(col);
+
+    if (row === this._rows) {
+      return;
+    }
+
+    this._board[row][col] = 0;
+  }
+
+  private _getFirstObjectInColumn(col: number): number {
+    // Find first obstruction from top down in column.
+    let row = 0;
+    while (row < this._rows && this._board[row][col] === 0) {
+      row++;
+    }
+    return row;
+  }
+
+  private _longestChainOnBoard(currentPlayer?: number): number {
+    return this._board.reduce((longestChainOnBoard, rowObj, row) => 
```

Comment:
This variable name is super long lmao, maybe `longestChain`

Reply:
We have a `longestChainOnBoard` and a `longestChainAtLocation`.

For the sake of people extending this class I feel like having the distinction is important no?

Reply:
Fair enough